### PR TITLE
MIR building: Stop using `unpack!` for `BlockAnd<()>`

### PR DESCRIPTION
**Pull request by Zalathar**
This is a subset of #127416, containing only the parts related to `BlockAnd<()>`.

The first patch removes the non-assigning form of the `unpack!` macro, because it is frustratingly inconsistent with the main form. We can replace it with an ordinary method that discards the `()` and returns the block.

The second patch then finds all of the remaining code that was using `unpack!` with `BlockAnd<()>`, and updates it to use that new method instead.

---

Changes since original review of #127416:
- Renamed `fn unpack_block` → `fn into_block`
- Removed `fn unpack_discard`, replacing it with `let _: BlockAnd<()> = ...` (2 occurrences)
- Tweaked `arm_end_blocks` to unpack earlier and build `Vec<BasicBlock>` instead of `Vec<BlockAnd<()>>`

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/mod.rs [hunk 1]
```diff
@@ -403,6 +403,15 @@ enum NeedsTemporary {
 #[must_use = "if you don't use one of these results, you're leaving a dangling edge"]
 struct BlockAnd<T>(BasicBlock, T);
 
+impl BlockAnd<()> {
+    /// Unpacks `BlockAnd<()>` into a [`BasicBlock`].
+    #[must_use]
+    fn unpack_block_and_unit(self) -> BasicBlock {
```
Comment (fmease) [inline compiler/rustc_mir_build/src/build/mod.rs:409]: This just bikeshedding but we could orientate ourselves by [`Result::into_ok`](https://doc.rust-lang.org/nightly/std/result/enum.Result.html#method.into_ok) and go with `into_block`. `Result::into_ok` is similar to this method in the sense that it "discharges" a "meaningless piece of information" from a more complex thing (`into_ok`: sum type w/ `!`, `into_block`: product type w/ `(=`). And doesn't mention `never` in its name.

**Review by fmease — APPROVED** [review]
I have a suggestion which you don't necessarily need to take to heart.

r=fmease,Nadrieril if Nadrieril approves as well

File: compiler/rustc_mir_build/src/build/mod.rs (hunk 1 above)
Comment (Zalathar) [reply compiler/rustc_mir_build/src/build/mod.rs:409]: Yeah, this sounds like an improvement to me. I'll try it out.

**Comment by Zalathar** [comment]
Renamed `fn unpack_block_and_unit` → `fn into_block` [(diff)](https://github.com/rust-lang/rust/compare/5066e6c169c06e31dc390ff8613ea07389955ee2..1cf4eb2ad258281fcf59ec93f41b43117bb5b824).

**Comment by Nadrieril** [comment]
Whoops I hadn't seen that you were waiting for me. LGTM, thx for merging 👍

**Pull request merged**